Align JavaScript SDK with existing Python SDK

peteski22 opened the pull request:
### Summary

Attempt to provide like-for-like functionality with a new JavaScript SDK API, based on the existing Python SDK patterns.

### API

* `client.listServers()` - List all servers
* `client.getServerHealth(name?)` - Get health for one or all servers
* `client.isServerHealthy(name)` - Check if server is healthy
* `client.servers.foo.listTools()` - List tools for specific server (`foo`)
* `client.servers.foo.callTool(name, args)` - Dynamic tool call (`bar`) for server (`foo`)
* `client.servers.foo.hasTool(name)` - Returns whether this server has the specified tool
* `client.servers.foo.tools.bar(args)` - Call tool (`bar`) for server (`foo`)
* `client.getToolSchemas(options?)` - Returns tool schemas from all servers with names transformed to `serverName__toolName` format. Useful for MCP servers aggregating/re-exposing tools from multiple upstream
  servers
* `client.getAgentTools(options?)` - Get callable functions for AI agents

All changes verified with linting, type checking, tests, and builds.

```bash
npm run lint
npm run format
npm run test
```

See: https://github.com/mozilla-ai/mcpd-sdk-python

peteski22 committed:
Add comprehensive .gitignore

* Exclude node_modules, build artifacts, and development files
* Include standard Node.js and TypeScript ignore patterns
* Prevent accidental commits of generated content

peteski22 committed:
Add JavaScript naming convention methods and minimize public API

* Add JavaScript-style method aliases to McpdClient:
  * getServers() -> servers()
  * getTools() -> tools()
  * getServerHealth() -> serverHealth()
  * getAgentTools() -> agentTools()
* Remove internal class exports (DynamicCaller, ServerProxy, FunctionBuilder, AgentFunction)
* Maintain full backward compatibility with Python-style methods
* Align public API surface with Python SDK conventions

peteski22 committed:
Update tests to use JavaScript naming conventions

* Replace all Python-style method calls with JavaScript equivalents
* Maintain comprehensive test coverage (41 tests passing)
* Verify backward compatibility through aliasing approach

peteski22 committed:
Update examples and documentation for JavaScript naming

* Update all examples to demonstrate JavaScript-style method calls
* Revise README.md API documentation to show new method names
* Update code snippets and usage patterns throughout documentation
* Maintain backward compatibility notes where appropriate

peteski22 committed:
Add project infrastructure and documentation

* Add comprehensive project documentation (SECURITY, CODE_OF_CONDUCT, CONTRIBUTING)
* Ensure proper open source project structure
* Provide clear guidelines for contributors and security reporting

khaledosman commented on examples/basic/index.js:
```diff
@@ -0,0 +1,98 @@
+    // Get health status for all servers
+    console.log('Server health:');
+    const health = await client.getServerHealth();
+    for (const [serverName, serverHealth] of Object.entries(health)) {
```
nit: I'm assuming this returns an object where the keys are the server names and the values are an object containing the health status? if so we could rename it to something that encompasses that better like healthResponseByServer / healthGroupedByServer or serverHealthsMap or similar


```suggestion
    const healthByServer = await client.getServerHealth();
    for (const [serverName, serverHealth] of Object.entries(healthByServer)) {
```

khaledosman commented on examples/basic/index.js:
```diff
@@ -0,0 +1,98 @@
+      console.log();
+
+      // Check if get_current_time tool exists
+      if (await client.hasTool('time', 'get_current_time')) {
```
nit: this doesn't seem very clear because its calling `client.hasTool` when clients don't really have tools, it also expects users to know about both the server name and the tool name so I'm not sure what its usecase is; alas I would rename it to something like `client.hasServerWithTool` or similar, wdyt?

khaledosman commented on examples/basic/index.js:
```diff
@@ -0,0 +1,98 @@
+
+        const nyTime = await client.call.time.get_current_time({ timezone: 'America/New_York' });
+        console.log('  New York:', nyTime);
+      }
```
`.call` is a reserved method on functions in javascript so it can be confusing https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/call

Is there really a need for this intermediary `call` object? how about calling it directly with `client.time` or `client.servers.time` instead? Would be nice to have the tool calls as camelCase aswell

khaledosman commented on examples/basic/index.js:
```diff
@@ -0,0 +1,98 @@
+
+    // Demonstrate error handling
+    console.log('\nError handling example:');
+    try {
+      // Try to call a non-existent tool
+      await client.call.nonexistent_server.nonexistent_tool();
+    } catch (error) {
+      if (error instanceof McpdError) {
+        console.log(`  Caught expected error: ${error.name} - ${error.message}`);
+      } else {
+        throw error;
+      }
+    }
+
+  } catch (error) {
+    if (error instanceof McpdError) {
+      console.error('mcpd error:', error.message);
+
+      // Show specific error handling
+      if (error.name === 'ConnectionError') {
+        console.error('Cannot connect to mcpd daemon. Is it running?');
+        console.error('Start it with: mcpd start');
+      } else if (error.name === 'AuthenticationError') {
+        console.error('Authentication failed. Check your API key.');
+      }
+    } else {
+      console.error('Unexpected error:', error);
+    }
+    process.exit(1);
+  }
```
nested try/catches can be difficult to follow through, I would rather consolidate them into a single try/catch with conditional logic based on the error type accordingly


```suggestion
  try {
    // Try to call a non-existent tool
    await client.call.nonexistent_server.nonexistent_tool();
  } catch (error) {
    if (error instanceof McpdError) {
      console.error('mcpd error:', error.message);

      // Show specific error handling
      if (error.name === 'ConnectionError') {
        console.error('Cannot connect to mcpd daemon. Is it running?');
        console.error('Start it with: mcpd start');
      } else if (error.name === 'AuthenticationError') {
        console.error('Authentication failed. Check your API key.');
      }
    } else {
      console.error('Unexpected error:', error);
    }
    process.exit(1)
  }
```

khaledosman commented on examples/basic/index.js:
```diff
@@ -0,0 +1,98 @@
+// Run the example
+if (require.main === module) {
+  main().catch(console.error);
+}
```
```suggestion
  main().catch(console.error);
```

khaledosman commented on examples/typescript/index.ts:
There should be no difference code-wise between the javascript and typescript examples, so I would delete this file or rely on either one alone

khaledosman commented on src/utils/cache.ts:
```diff
@@ -0,0 +1,88 @@
+ * Cache decorator for async methods.
+ * Caches successful results and certain error types.
+ */
+export function cached<T extends (...args: any[]) => Promise<any>>(
```
decorators are not yet natively part of the language.. considering we're calling it as a function that takes a function I would rename it as a verb, how about `memoize`?

khaledosman commented on src/utils/typeConverter.ts:
```diff
@@ -0,0 +1,140 @@
+ * and properly maps nullable types to TypeScript's type system.
+ *
+ * Used primarily by the FunctionBuilder to create accurate type information
+ * for dynamically generated functions.
```
I believe this is converting mcpd parameter types into javascript language types and not "json/schema types into typescript ones", typescript types are only checked at build/compilation time, so they aren't really part of any business or runtime logic so I would rename things accordingly

khaledosman commented on src/client.ts:
```diff
@@ -0,0 +1,612 @@
+    ServerNotFoundError,
+    ServerUnhealthyError,
+    AuthenticationError,
+  ]);
```
nit: `private` in typescript doesn't really mean private see https://stackoverflow.com/questions/12713659/why-can-i-access-typescript-private-members-when-i-shouldnt-be-able-to 

if you really want to make these private you can prefix the variable/method names with `#`  see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_elements

khaledosman commented on src/client.ts:
````diff
@@ -0,0 +1,612 @@
+   * console.log(servers); // ['time', 'fetch', 'git']
+   * ```
+   */
+  async servers(): Promise<string[]> {
````
nit: would be nice to have function methods include a verb like `getServers` or `fetchServes` or similar, but I understand if we want to keep it consistent with the python sdk version

khaledosman commented on src/dynamicCaller.ts:
```diff
@@ -0,0 +1,121 @@
+    this.client = client;
+
+    // Return a Proxy to intercept property access
+    return new Proxy(this, {
+      get: (target, serverName: string | symbol) => {
+        if (typeof serverName !== 'string') {
+          return undefined;
+        }
+        return new ServerProxy(target.client, serverName);
+      },
+    });
+  }
```
this looks unnecessarily complicated, do we really need to use a proxy?

khaledosman commented on src/dynamicCaller.ts:
```diff
@@ -0,0 +1,121 @@
+   * @param serverName - The name of the MCP server this proxy represents
+   */
+  constructor(client: McpdClient, serverName: string) {
+    this.client = client;
+    this.serverName = serverName;
+
+    // Return a Proxy to intercept method calls
+    return new Proxy(this, {
+      get: (target, toolName: string | symbol) => {
+        if (typeof toolName !== 'string') {
+          return undefined;
+        }
+
+        // Return a function that will call the tool
+        return async (args?: Record<string, any>) => {
+          // Check if the tool exists
+          const hasTool = await target.client.hasTool(target.serverName, toolName);
+          if (!hasTool) {
+            throw new ToolNotFoundError(
+              `Tool '${toolName}' not found on server '${target.serverName}'. ` +
+                `Use client.tools('${target.serverName}') to see available tools.`,
+              target.serverName,
+              toolName
+            );
+          }
+
+          // Perform the tool call
+          return target.client._performCall(target.serverName, toolName, args);
+        };
+      },
+    });
+  }
```
same here, I'm not sure it adds much value compared to directly calling the corresponding methods or having it directly accessible without a proxy, whats the idea behind having a caller or a `.call` proxy on the client as opposed to just calling/finding the server and/or its tools accordingly via the other methods?

khaledosman commented on package.json:
```diff
@@ -0,0 +1,72 @@
+{
+  "name": "@mozilla-ai/mcpd",
+  "version": "0.1.0",
+  "description": "JavaScript/TypeScript SDK for interacting with mcpd",
```
I think we need to add a `"type": "module"` to support es6 imports https://stackoverflow.com/questions/61401475/why-is-type-module-in-package-json-file

peteski22 committed:
Add core dependencies and API path infrastructure

- Add zod dependency for JSON Schema to Zod conversion
- Add centralized API path constants in apiPaths.ts
- Update package-lock.json for new dependencies

This establishes the foundation for dual framework compatibility
and cleaner API path management in the SDK.

peteski22 committed:
Add dual framework support to function builder

- Update AgentFunction interface for LangChain JS and Vercel AI SDK compatibility
- Add JSON Schema to Zod conversion with proper type handling
- Use .nullable().optional() for OpenAI structured outputs compatibility
- Fix empty schema handling to return z.object({}) instead of errors
- Add invoke/execute methods for framework-specific execution patterns
- Replace __name__/__doc__ with standard name/description properties

This enables generated functions to work directly with both AI frameworks
without requiring conversion layers.

peteski22 committed:
Implement unified getAgentTools API with format support

- Add ToolFormat type ('array' | 'object' | 'map') for cross-framework compatibility
- Add TypeScript overloads for getAgentTools() with proper return type inference
- Replace framework-specific methods with single data structure-based approach
- Update API path handling to use centralized apiPaths constants
- Align tools() method behavior with Python SDK patterns
- Improve tool call error handling and response parsing
- Fix tool call payload format to match mcpd daemon expectations

This delivers the unified developer experience where getAgentTools()
returns objects that work directly with both LangChain JS and Vercel AI SDK
without requiring conversion functions.

peteski22 committed:
Updated README

peteski22 committed:
Add LangChain example

peteski22 committed:
Add Vercel-AI example

peteski22 committed:
Update examples/basic/index.js

Co-authored-by: Khaled Osman <thisiskhaled@proton.me>

peteski22 committed:
Addressing PR feedback

SDK Architecture Changes

* Replace .call namespace with .servers namespace to avoid JavaScript reserved word collision
* Add client.getServer(name) method for programmatic server access
* Add hasTool(name) method on server proxies to check tool existence
* Support both snake_case (MCP spec) and camelCase (JavaScript idiom) for tool names with automatic conversion
* Replace TypeScript private with JavaScript # for true runtime privacy
* Split cache TTL into healthCacheTtl (10s default) and serverCacheTtl (60s default) for granular control
* Implement unified getAgentTools() API with format parameter ('array', 'object', 'map') for framework compatibility

API Response Handling

* Fix getServers() to handle array response format (not wrapped in object)
* Fix getServerHealth() to transform array response into Record format
* Fix _performCall() to parse JSON strings returned directly by API
* Add ErrorModel and ErrorDetail types for structured API error responses
* Update error handling to parse and format ErrorModel with error details
* Update ToolExecutionError to use ErrorModel instead of unknown for typed error details
* Make ErrorDetail.location and ErrorDetail.value optional fields

Type Safety Improvements

* Replace any with unknown throughout codebase for data flow (parameters, return values)
* Keep any only for framework types (Zod schemas, JSON Schema) with eslint-disable comments
* Export ErrorDetail and ErrorModel types from public API
* Add proper type casting where needed (e.g., params as Record<string, unknown>)

Code Cleanup

* Remove unused cached decorator (47 lines)
* Remove unused DynamicCaller class
* Remove unused RequestOptions interface
* Remove ServersResponse and ToolCallResponse from exports (internal only)
* Remove unused math tool example code
* Remove redundant typescript example directory
* Remove redundant with-agents example directory
* Remove default export to eliminate mixed exports build warning

Tooling & Infrastructure

* Upgrade ESLint 8→9 to support flat config
* Install jiti and ts-node for TypeScript config loading
* Add ignores for dist/, coverage/, node_modules/ in ESLint config
* Remove deprecated --ext flags from lint scripts
* Rename vite.config.ts to vite.config.mts to fix CJS deprecation warning
* Fix examples to use ES module syntax (import, not require)

Documentation & Tests

* Update all examples from .call.* to .servers.*
* Add JS/TS clarification note to README
* Update all tests to match actual API behavior (ErrorModel, JSON strings)
* Fix test mocks to use realistic data (IANA timezones, actual API response formats)

khaledosman commented on examples/basic/index.js:
```diff
@@ -0,0 +1,96 @@
+        console.log('Tokyo:');
+        console.log(tokyoTime);
+
+        const nyTime = await client.servers.time.get_current_time({ timezone: 'America/New_York' });
```
do these need to be in snake_case?

khaledosman commented on examples/basic/index.js:
```diff
@@ -0,0 +1,96 @@
+      const tools = await client.getTools('time');
+      for (const tool of tools) {
+        console.log(`  - ${tool.name}: ${tool.description || 'No description'}`);
+      }
```
does this get all tools with that name across all servers? can there for example be duplicates across multiple servers?

peteski22 replied:
It's been updated to:

```javascript
const tools = await client.getToolSchemas();
```

You can supply an array of server names to filter it down too.

It returns the schemas with the format `serverName__toolName` to prevent collisions (captured this in the function documentation).

This is useful for representing *everything* together, but usually I would expect end users get tools for a particular server (e.g. named `foo`):

```javascript
const tools = await client.servers.foo.listTools()
```

khaledosman commented on examples/langchain/index.js:
```diff
@@ -0,0 +1,127 @@
+      const cleanDescription = tool.description
+        .split('\n')[0]  // Take only first line (actual newline)
+        .replace(/\s+/g, ' ')  // Replace multiple spaces with single space
+        .trim();
```
is this something that should be done within the library?

peteski22 replied:
It was just to get a small bit of the desc for display, I changed it to only take the first line before a line break.

khaledosman commented on src/errors.ts:
```diff
@@ -0,0 +1,188 @@
+    this.name = 'TimeoutError';
+    this.operation = operation;
+    this.timeout = timeout;
+    Object.setPrototypeOf(this, TimeoutError.prototype);
```
What's the reason behind manually setting the prototype of the object for these custom errors? I don't think its needed.

I think you need `Error.captureStackTrace(this, this.constructor);` for all these custom errors aswell to maintain the stack trace see https://futurestud.io/tutorials/node-js-create-your-custom-error & https://dev.to/apisurfer/from-custom-error-types-to-a-custom-error-subtype-3nkf

peteski22 committed:
Implement dependency injection for private methods

* Add PerformCallFn and GetToolsFn types for dependency injection
* Refactor proxy classes to accept injected functions via constructor
* Update FunctionBuilder to use injected PerformCallFn
* Make client._performCall truly private with # syntax
* Inject bound functions from McpdClient to all dependencies

Verified with npm run lint, test, and build

peteski22 committed:
Update tests for dependency injection architecture

* Update functionBuilder tests to mock injected PerformCallFn
* Update client tests to remove _performCall tests
* Remove unused McpdError import

Verified with npm run lint, test, and build

peteski22 committed:
Update examples for new API patterns

* Update basic example to use .tools namespace
* Update langchain example to use getAgentTools options object
* Update vercel-ai example to use getAgentTools options object

Verified with npm run lint, test, and build

peteski22 committed:
Update documentation for new API patterns

* Update README with .tools namespace examples
* Update langchain README with new API syntax
* Update vercel-ai README with new API syntax
* Update README to show getAgentTools options object

Verified with npm run lint, format, test, and build

peteski22 committed:
Fix test script to exit after run

* Change test script from vitest to vitest run
* Add test:watch script for development

Verified with npm run lint, test, and build

peteski22 committed:
Apply code formatting and linting fixes

* Run prettier formatting on markdown and TypeScript files
* Fix ESLint issues across codebase
* Update import formatting and style consistency

Verified with npm run lint, format, test, and build

peteski22 committed:
Update Node version requirement to v22 LTS

* Update package.json engines to require Node >=22.0.0
* Update GitHub publish action to use Node 22
* Update example READMEs to specify Node 22 LTS or higher
* Update tsconfig.json to target ES2023 (matches Node 22 capabilities)

Verified with npm run lint, test, and build

peteski22 committed:
Add CI workflows for tests and linting

* Add tests.yaml workflow to run unit tests and build
* Add lint.yaml workflow to check linting and formatting
* Update publish.yml with workflow_dispatch and tag checkout

Verified with npm run lint, format, test, and build

peteski22 committed:
Add TypeScript checking to linting and CI

* Add tsconfig.test.json to enable TypeScript checking for test files
* Add typecheck:tests and typecheck:all scripts to package.json
* Add function overloads to getAgentTools for proper type narrowing
* Update lint script to include typecheck:all
* Update lint.yaml workflow to run typechecking
* Fix test files to properly handle array indexing with noUncheckedIndexedAccess
* Fix __name__ → .name in functionBuilder test

Verified with npm run lint, format, test, and build

peteski22 committed:
Improve error classes with Error.captureStackTrace

* Replace Object.setPrototypeOf with Error.captureStackTrace
* Maintains proper stack traces in Node.js
* Removes ES5 workaround no longer needed with ES2023 target
* Addresses PR review feedback

Verified with npm run lint, test, and build

peteski22 committed:
Align API method names with specification

* Rename getServers() to listServers()
* Remove serverHealth() wrapper, use getServerHealth() directly
* Make getTools() private as #getToolsByServer() - users should use client.servers.foo.listTools()
* Update all tests, documentation, and examples
* Add health check mocks to agentTools test

API now matches spec:
- client.listServers()
- client.getServerHealth(name?)
- client.isServerHealthy(name)
- client.servers.foo.listTools()
- client.servers.foo.tools.bar(args)
- client.getAgentTools(options?)

Verified with npm run lint, test, and build

peteski22 committed:
Add getToolSchemas() for MCP server tool aggregation

* Add client.getToolSchemas(options?) method
* Transforms tool names to serverName__toolName format to prevent clashes
* Supports filtering by servers via options.servers
* Useful for MCP servers aggregating/re-exposing upstream tools
* Add comprehensive tests (all servers, filtered, empty, error handling)
* Update README with clear documentation about name transformation

Verified with npm run lint, format, test, and build

peteski22 committed:
Fix config and examples

* Rename McpdSDK to McpdSdk in vite config
* Move hasTool() and callTool() out of .tools namespace in examples
* Handle optional description field in langchain example

peteski22 committed:
Add comprehensive API surface tests

* Add tests to cover the public API surface and access patterns
* Prevent regressions by explicitly testing each documented method

peteski22 committed:
Align SDK types with MCP Protocol Specification

* Update Tool interface to match MCP spec (2025-06-18)
* Make inputSchema required and description optional per protocol
* Remove ToolSchema type in favor of canonical Tool type
* Add ToolAnnotations interface for tool behavior hints
* Update ServerHealth structure with required name and status fields
* Move callTool and hasTool methods from .tools namespace to Server class
* Update all type references and tests
* Add comprehensive JSDoc comments linking to spec

peteski22 committed:
Optimize tool fetching with health pre-filtering and parallelization

* Pre-filter unhealthy servers before fetching tools
* Parallelize tool fetches across healthy servers with Promise.allSettled
* Cache individual server health from bulk health checks
* Update comprehensive documentation for getToolSchemas and agentTools
* Update #getToolsByServer documentation with all thrown errors
* Update tests to match new parallel execution pattern

khaledosman commented on package.json:
```diff
@@ -0,0 +1,82 @@
+  },
+  "homepage": "https://github.com/mozilla-ai/mcpd-sdk-javascript#readme",
+  "files": [
+    "dist",
+    "src"
+  ],
```
typically you don't need "src" to be in the published package since `dist/` should include both the compiled javascript code as well as the generated type definition "d.ts" files. The idea is to only publish the files needed to make the package as small is possible, this is nice for faster download times, but also more importantly translates into performance/page-load times due to downloading more stuff, although typically most modern bundlers have tree-shaking nowadays to eliminate dead code.

The "main" property of the package.json points to "dist/index.js" which is the entry file that will be read whenever someone tries to import something from the library

Alternatively we can also use an `.npmignore` to determine which files should not be included in the published package as opposed to the "files" property in package.json

khaledosman commented on package.json:
```diff
@@ -0,0 +1,82 @@
+  "exports": {
+    ".": {
+      "types": "./dist/index.d.ts",
+      "require": "./dist/index.js",
+      "import": "./dist/index.mjs"
```
This is not bad to provide support for old nodejs versions, but no longer necessary, since the current active LTS version of nodejs is v22 which already supports "type": "import" and es6 syntax https://nodejs.org/en/about/previous-releases

peteski22 committed:
Config updates

- Remove "type": "module" from package.json to follow industry standard (Stripe, Anthropic, Vercel AI)
  - Add ssr: true to vite config for proper Node.js API support (process.env, fetch)
  - Bundle lru-cache dependency for simpler installation
  - Update target to node22 for consistency

khaledosman commented on package.json:
```diff
@@ -1,7 +1,6 @@
 {
   "name": "@mozilla-ai/mcpd",
   "version": "0.1.0",
-  "type": "module",
```
I believe this is still required for es modules support

peteski22 committed:
Remove src from published package files

* Only include dist/ in published package
* Source TypeScript files not needed by consumers
* All entry points reference dist/ not src/
* Reduces package size for faster downloads
